src/System.Text.Json/src/System/Text/Json/Serialization/ReadStackFrame.cs

```diff
@@ -41,35 +44,56 @@ internal struct ReadStackFrame
+        }
 
-        public bool IsCollectionForClass => IsEnumerable || IsDictionary || IsIDictionaryConstructible;
-        public bool IsCollectionForProperty => IsEnumerableProperty || IsDictionaryProperty || IsIDictionaryConstructibleProperty;
+        public bool IsProcessingCollectionForProperty()
+        {
+            return IsProcessingProperty(ClassType.Enumerable | ClassType.Dictionary | ClassType.IDictionaryConstructible);
+        }
 
-        public bool IsIDictionaryConstructible => JsonClassInfo.ClassType == ClassType.IDictionaryConstructible;
-        public bool IsDictionary => JsonClassInfo.ClassType == ClassType.Dictionary;
+        public bool IsProcessingEnumerableOrDictionary()
+        {
+            return IsProcessing(ClassType.Enumerable | ClassType.Dictionary | ClassType.IDictionaryConstructible) ||
+                IsProcessingProperty(ClassType.Enumerable | ClassType.Dictionary | ClassType.IDictionaryConstructible);
+        }
 
-        public bool IsDictionaryProperty => JsonPropertyInfo != null &&
-            !JsonPropertyInfo.IsPropertyPolicy &&
-            JsonPropertyInfo.ClassType == ClassType.Dictionary;
-        public bool IsIDictionaryConstructibleProperty => JsonPropertyInfo != null &&
-            !JsonPropertyInfo.IsPropertyPolicy && (JsonPropertyInfo.ClassType == ClassType.IDictionaryConstructible);
+        public bool IsProcessingDictionary()
+        {
+            return IsProcessing(ClassType.Dictionary) || IsProcessingProperty(ClassType.Dictionary);
+        }
 
-        public bool IsEnumerable => JsonClassInfo.ClassType == ClassType.Enumerable;
+        public bool IsProcessingIDictionaryConstructible()
+        {
+            return IsProcessing(ClassType.IDictionaryConstructible) || IsProcessingProperty(ClassType.IDictionaryConstructible);
+        }
 
-        public bool IsEnumerableProperty =>
-            JsonPropertyInfo != null &&
-            !JsonPropertyInfo.IsPropertyPolicy &&
-            JsonPropertyInfo.ClassType == ClassType.Enumerable;
+        public bool IsProcessingDictionaryOrIDictionaryConstructible()
+        {
+            return IsProcessing(ClassType.Dictionary | ClassType.IDictionaryConstructible) || IsProcessingProperty(ClassType.Dictionary | ClassType.IDictionaryConstructible);
+        }
 
-        public bool IsProcessingEnumerableOrDictionary => IsProcessingEnumerable || IsProcessingDictionary || IsProcessingIDictionaryConstructible;
-        public bool IsProcessingDictionary => IsDictionary || IsDictionaryProperty;
-        public bool IsProcessingIDictionaryConstructible => IsIDictionaryConstructible || IsIDictionaryConstructibleProperty;
-        public bool IsProcessingEnumerable => IsEnumerable || IsEnumerableProperty;
+        public bool IsProcessingEnumerable()
+        {
+            return IsProcessing(ClassType.Enumerable) || IsProcessingProperty(ClassType.Enumerable);
+        }
+
+        public bool IsProcessing(ClassType classTypes)
```

Comment:
Can you document this helper and explain the logic a bit with a comment?

Reply:
I will document each method, and normalize the naming across them so it is clear whether they apply to an object, property or both.